### ISSUE TITLE
get multiple json files

### ISSUE THREAD

**Issue by nousacademy** [aws/aws-sdk-js#623]
Is there a way to display all contents of a folder, in my case I have json files, and render them all?? Instead of one at a time??

**Comment by lsegal** [comment]
Can you provide more information about what you are trying to do, which version of the SDK you are using and environment (browser / Node.js)? Code would also be helpful to understand and reproduce your issue.

**Comment by nousacademy** [comment]
I'm using the NodeJS environment, version 2.1.14.  I want to list all objects within a folder which I've done, but then GET all my json files within it, to display them. So far I see that I can only GET one json file at a time.

**Comment by lsegal** [comment]
@nousacademy if you are referring to the `AWS.S3` service, then you are correct that you can only GET one single object from S3 at a time. Note that Node.js will pool and re-use connections (especially in Node 0.12+), so these requests may end up re-using the same HTTP connection.

**Comment by AdityaManohar** [comment]
To add to what @lsegal said, you can use the [`async`](https://github.com/caolan/async) module with the SDK to download your Amazon S3 objects serially or in parallel. For example:

```javascript
var AWS = require('aws-sdk');
var async = require('async');

var s3 = new AWS.S3({Bucket: 'foo'});

// Called after all iterators complete
// Write your clean up here
var callback = function(err) {}

s3.listObject(params, function(err, data){
  // GET objects in parallel
  async.each(data.Contents, function(key, cb){
    var stream = s3.getObject({Key: key}).createReadStream();
    stream.on('end', cb);
    var file = fs.createWriteStream('/path/to/file');
    stream.pipe(file);
  }, callback);
});
```

Hope this helps!

**Comment by nousacademy** [comment]
@lsegal @AdityaManohar thanks, I'm testing it out now!

**Comment by nousacademy** [comment]
I'm getting this error from aws:

    /Users/nousacademy/Desktop/Projects/testapp/node_modules/aws-sdk/lib/request.js:32
              throw err;
                    ^
    Error: Can't set headers after they are sent.

**Comment by AdityaManohar** [comment]
Can you post the code you are using?

**Comment by nousacademy** [comment]
s3.listObjects(params, function(err, data) {
            res.send(data);

        async.each(data.Contents, function(key, cb) {

            var params2 = {
                Bucket: 'johnsbookmarkapp',
                Key: key.Key
            }
            s3.getObject(params2, function(err, data) {

                res.send(data.Body.toString())
            });


            //console.log(key.Key)
        });
    });

Basically wanted to get the list, then get all the objects, without having to write to file

**Comment by AdityaManohar** [comment]
Looks like you are trying to do `response.send()` twice. Here:

> `res.send(data);`

and here:

> `res.send(data.Body.toString())`

This effectively means you are sending more than one response per request which is not possible. See [Stack Overflow: cant-do-multiple-res-send-in-express-js](http://stackoverflow.com/questions/25542521/cant-do-multiple-res-send-in-express-js).

There are a couple ways to work around this:

1. Make a `AWS.S3.listObjects()` call from your client (which may be a browser or mobile application) and follow this up with multiple `AWS.S3.getObject()` calls. Amazon S3 supports CORS and you can make requests directly to Amazon S3 with the SDK in the browser. You can then render this data incrementally as each `AWS.S3.getObject()` request completes.

2. Alternatively, you can assemble all your JSON data before sending a response back to your client. For example:

```javascript
var payload = {};
payload.objects = [];

// Send your response after all requests have completed
var callback = function(err) {
  res.send(payload);
}

s3.listObjects(params, function(err, data){
  payload.list = data.Contents;
  // GET objects in parallel
  async.each(data.Contents, function(file, cb){
    var stream = s3.getObject({Key: file.Key}, function(err, data) {
      payload.objects.push(data.Body.toString());
      cb();
    })
  }, callback);
});
```
You can then parse this response data on your client.

3. In the third approach, you can use `response.write()` instead of `response.send()`. This will however, make it harder to delimit and parse the data on your client.

If you use approach (2) or (3), you will only have access to the data after all requests have completed. This means you will not be able to incrementally render your objects as they are fetched.

Also, irrespective of your approach, the number of requests you make to Amazon S3 does not change.

Hope this helps!

**Comment by nousacademy** [comment]
@AdityaManohar wow nice work around thanks!

**Comment by nousacademy** [comment]
@AdityaManohar It works now, but the order is destroyed. payload.list[0] is not equal to the contents associated with payload.objects[0]

**Comment by lsegal** [comment]
@nousacademy good point, you will want to assemble your data using the key as a key in a map instead of a list, which would keep order.

That said, I would strongly encourage you to follow the suggestion in (1) instead of (2). Using the second method imposes a lock around the entire UI while it waits to get *all* data. A more responsive UI would incrementally fetch items and populate them as they are made available, making the loading behavior much more seamless to the user. Specifically, you should list all objects, render the list, and then fetch each object's data in the background and slowly populate that data as it becomes available. By the time the user clicks on these objects, they will be loaded. You could even wait until a user clicks on an object to fetch the data, reducing the user's bandwidth usage, but that depends on how you display data in your UI. Either way, showing the list first will provide a "progressive loading" that will likely look much more natural to your design.

**Comment by nousacademy** [comment]
@lsegal I'm just curious as to how ill delete and update that object? Is the only way to do that using the absolute path??

**Comment by nousacademy** [comment]
I got it nevermind

    s3.getObject({Bucket: 'bucket', Key: 'key'}).on('success', function(response) {
      console.log("Key was", response.request.params.Key);
    }).send();

**Comment by nousacademy** [comment]
I cannot get the <code>response.request.data.Body</code> with <code>response.request.params.Key</code> it keeps giving me an undefined when I try to get the Body. What am I doing wrong????

**Comment by AdityaManohar** [comment]
The [`data`](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Response.html#data-property) property is on the [`Response`](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Response.html) object, not the [`Request`](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Request.html) object. You can access it as `response.data.Body`.

**Comment by nousacademy** [comment]
haha oops... <img src='http://images.sodahead.com/polls/002658747/1856902986_barack_obama_facepalm_xlarge.jpeg'>

**Comment by nousacademy** [comment]
Thanks @lsegal and @AdityaManohar I have everything! Appreciate the help!

**Comment by nousacademy** [comment]
Ok I thought I was done:

     for (var i = 0; i < jsonData.length; i++) {
 
                    $.get('/fldrContents', {
                        contents: jsonData[i].Key
                    }, function(data) {
                        var b = JSON.parse(data.body);
                  
                    
                    });
        }

I'm lost at how to put all the responses I get from my get call into one object. Theyre all coming individually. I tried to do this for the response with no luck: http://stackoverflow.com/questions/21819905/jquery-ajax-calls-in-for-loop

**Comment by nousacademy** [comment]
I got it I had to use a promise

**Comment by AdityaManohar** [comment]
@nousacademy I'm going to mark this issue as closed because it looks like you have this working. Feel free to reopen this issue or open another issue if you have other questions.

**Comment by nousacademy** [comment]
@AdityaManohar having another issue. When I try to GET one of my folders that has no subfolders and only files within it. It gets all of my files BUT it also gets the folder path which breaks my app. Example response:

    Object {path: "bookmarks/Bookmarks Toolbar/", body: ""}

As you see its getting the folder without any files. Because obviously a folder by itself is nothing! How do I get around this??

**Comment by lsegal** [comment]
@nousacademy if there is actually an S3 object named `bookmarks/Bookmarks Toolbar/` that has no contents, my suggestion would be to remove that object in S3, as it doesn't represent any actual data in your application. S3 is a flat key namespace, so the concept of "folders" does not exist and does not need to be represented specially with an object placeholder. 

That said, if you do have value in keeping those folders in your S3 bucket as real objects you could and just check for an ending "/" on the path to determine if it's a "folder" or a "file" in your app. Or you could add metadata to the object to be more explicit.

**Comment by nousacademy** [comment]
@lsegal got it
